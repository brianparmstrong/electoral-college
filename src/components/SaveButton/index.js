import React from 'react';
import { arrayOf, number } from 'prop-types';

const SaveButton = ({
  currentEVTotals,
  currentPVTotals
}) => {
  const saveStateStatus = () => {
    const states = document.getElementsByClassName('stateWrapper');
    const statesLength = states.length;
    let currentStateStatus = [];
    let currentPopVoteStatus = [];
    let state;
    let stateData;
    let pvData;

    const dynamicSort = property => {
      const sortOrder = 1;
      return (a,b) => {
        const result = (a[property] < b[property]) ? -1 : (a[property] > b[property]) ? 1 : 0;
        return result * sortOrder;
      };
    };

    const setStateStatus = stateData => {
      currentStateStatus.push(stateData);
      currentStateStatus.sort(dynamicSort('name'));

      return currentStateStatus;
    };

    const setPopVoteStatus = pvData => {
      let inputValues = [];
      let popVoteData = {};
      const inputElems = pvData?.elem?.querySelectorAll('.pvInput');
      if (inputElems) {
        for (let i=0; i<inputElems.length; i++) {
          inputValues[i] = inputElems[i].value;
        }
      }
      popVoteData = {name: pvData.stateName, values: inputValues};
      currentPopVoteStatus.push(popVoteData);
      currentPopVoteStatus.sort(dynamicSort('name'));

      return currentPopVoteStatus;
    };

    for (state = 0; state < statesLength; state++) {
      stateData = {
        evs: states[state].childNodes[0].getAttribute('data-evs'),
        name: states[state].childNodes[0].getAttribute('data-statename'),
        stateCode: states[state].childNodes[0].getAttribute('data-statecode'),
        winner: states[state].childNodes[0].getAttribute('data-winner')
      };
      pvData = {
        elem: states[state].parentElement.childNodes[1],
        name: states[state].childNodes[0].getAttribute('data-statename')
      };
      setStateStatus(stateData);
      setPopVoteStatus(pvData);
    }
    if (typeof(Storage) !== 'undefined') {
        localStorage.setItem('ElectoralCollegeStatus', JSON.stringify(currentStateStatus));
        localStorage.setItem('WinnerTakeAllTotals', JSON.stringify(currentEVTotals));
        localStorage.setItem('PopularVoteStatus', JSON.stringify(currentPopVoteStatus));
        localStorage.setItem('PopularVoteTotals', JSON.stringify(currentPVTotals));
        currentStateStatus = [];
        currentPopVoteStatus = [];
        alert('Saved!');
    } else {
      alert('Please enable Storage to save your selections.');
    }
  };
  return (
    <div className='buttonWrapper'>
      <button id='saveButton' type='button' onClick={saveStateStatus}>Save</button>
    </div>
  );
};

SaveButton.propTypes = {
  currentEVTotals: arrayOf(number).isRequired,
  currentPVTotals: arrayOf(number).isRequired
};

export default SaveButton;
