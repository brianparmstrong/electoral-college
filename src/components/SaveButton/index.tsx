import { IfcSaveButton, PopVotesData } from '../../types';

type PvData = {
  elem: HTMLElement;
  name: string;
};

type StateData = {
  evs: string;
  name: string;
  stateCode: string;
  winner: string;
};

const SaveButton = (saveButtonProps: IfcSaveButton) => {
  const { currentEVTotals, currentPVTotals, handleSave } = saveButtonProps;
  const saveStateStatus = () => {
    const states = document.getElementsByClassName('stateWrapper');
    const statesLength = states.length;
    let currentStateStatus: StateData[] = [];
    let currentPopVoteStatus: PopVotesData[] = [];
    let stateData: StateData;
    let pvData: PvData;

    const dynamicSort = (property: string) => {
      const sortOrder = 1;
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore for `a` and `b`
      return (a, b) => {
        const result =
          a[property] < b[property] ? -1 : a[property] > b[property] ? 1 : 0;
        return result * sortOrder;
      };
    };

    const setStateStatus = (stateData: StateData) => {
      currentStateStatus.push(stateData);
      currentStateStatus.sort(dynamicSort('name'));

      return currentStateStatus;
    };

    const setPopVoteStatus = (pvData: PvData) => {
      const inputValues: string[] = [];
      const popVoteData = { name: '', values: inputValues };
      const inputElems = pvData?.elem?.querySelectorAll(
        '.pvInput'
      ) as NodeListOf<HTMLInputElement>;
      if (inputElems) {
        for (let i = 0; i < inputElems.length; i++) {
          inputValues[i] = inputElems[i].value;
        }
      }
      popVoteData.name = pvData.name;
      popVoteData.values = inputValues;
      currentPopVoteStatus.push(popVoteData);
      currentPopVoteStatus.sort(dynamicSort('name'));

      return currentPopVoteStatus;
    };

    for (let state = 0; state < statesLength; state++) {
      const stateNode = states[state].childNodes[0] as HTMLDivElement;
      stateData = {
        evs: stateNode.getAttribute('data-evs') as string,
        name: stateNode.getAttribute('data-statename') as string,
        stateCode: stateNode.getAttribute('data-statecode') as string,
        winner: stateNode.getAttribute('data-winner') as string,
      };
      pvData = {
        elem: states[state].parentElement?.childNodes[1] as HTMLDivElement,
        name: stateNode.getAttribute('data-statename') as string,
      };
      setStateStatus(stateData);
      setPopVoteStatus(pvData);
    }

    localStorage.setItem(
      'ElectoralCollegeStatus',
      JSON.stringify(currentStateStatus)
    );
    localStorage.setItem(
      'WinnerTakeAllTotals',
      JSON.stringify(currentEVTotals)
    );
    localStorage.setItem(
      'PopularVoteStatus',
      JSON.stringify(currentPopVoteStatus)
    );
    localStorage.setItem('PopularVoteTotals', JSON.stringify(currentPVTotals));
    currentStateStatus = [];
    currentPopVoteStatus = [];
    handleSave();
  };
  return (
    <div className="buttonWrapper">
      <button id="saveButton" type="button" onClick={saveStateStatus}>
        Save
      </button>
    </div>
  );
};

export default SaveButton;
