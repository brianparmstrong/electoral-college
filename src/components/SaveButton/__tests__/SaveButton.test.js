import { render } from '@testing-library/react';
import SaveButton from '../';

const MOCK_TOTALS = {
  currentEVTotals: [280, 210, 0, 0, 0],
  currentPVTotals: [255, 230, 3, 1, 1],
};

it('renders correctly', () => {
  const { container } = render(<SaveButton {...MOCK_TOTALS} />);
  expect(container.firstChild).toMatchSnapshot();
});
