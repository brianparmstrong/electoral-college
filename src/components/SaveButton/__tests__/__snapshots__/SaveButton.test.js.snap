// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  class="buttonWrapper"
>
  <button
    id="saveButton"
    type="button"
  >
    Save
  </button>
</div>
`;
